### PR TITLE
Improve symbol visibility handling for dynamic libraries.

### PR DESCRIPTION
**Pull request by michaelwoerister**
This will hopefully fix issue https://github.com/rust-lang/rust/issues/37530 and maybe also https://github.com/rust-lang/rust/issues/32887.

I'm relying on @m4b to post some numbers on how awesome the improvement for cdylibs is `:)`

cc @rust-lang/compiler @rust-lang/tools @cuviper @froydnj
r? @alexcrichton

### CODE REVIEW

File: src/test/run-make/symbol-visibility/Makefile [hunk 1]
```diff
@@ -0,0 +1,26 @@
+include ../tools.mk
+
+all:
```
Comment (alexcrichton) [inline src/test/run-make/symbol-visibility/Makefile]: This'll probably need a guard or two for only running on Linux (I think). 

Either that or it'll need some tweaks to run on OSX/Windows

File: src/librustc_trans/back/symbol_export.rs [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    }
+}
+
+pub fn crate_export_threshold(crate_type: config::CrateType)
+                                     -> SymbolExportLevel {
+    match crate_type {
+        config::CrateTypeExecutable |
+        config::CrateTypeStaticlib  |
+        config::CrateTypeCdylib     => SymbolExportLevel::C,
+        config::CrateTypeProcMacro  |
```
Comment (alexcrichton) [inline src/librustc_trans/back/symbol_export.rs]: I think we can actually switch this to a c export level, it's essentially like an executable where nothing else will ever link to it again.

**Review by alexcrichton — COMMENTED** [review]
Looks great to me!

Could you also post a small tl;dr; of the high-level changes here? I got lost a bit in the refactorings, but it seemed like it was (a) a tweak to how we call the linker and (b) a tweak to sometimes set symbols to hidden. Clarifications on these two though would be welcome :)

Also, I'm curious about the rlib-into-dylib case as well. Presumably rlibs don't have hidden symbols if they're public exported, right? In that case, if it's linked into a cdylib we're using the "fancy linker argument" to not export the symbols? If it's linked into an executable or goes into a staticlib the symbol will still be exported? (unsure on this last bit)

File: src/librustc_trans/back/lto.rs [hunk 3]
```diff
@@ -38,17 +55,31 @@ pub fn run(sess: &session::Session, llmod: ModuleRef,
...
     // Make sure we actually can run LTO
     for crate_type in sess.crate_types.borrow().iter() {
-        match *crate_type {
-            config::CrateTypeExecutable |
-            config::CrateTypeCdylib |
-            config::CrateTypeStaticlib => {}
-            _ => {
-                sess.fatal("lto can only be run for executables and \
+        if !crate_type_allows_lto(*crate_type) {
+            sess.fatal("lto can only be run for executables and \
```
Comment (nagisa) [inline src/librustc_trans/back/lto.rs]: This message does not match the `crate_type_allows_lto` function above anymore.

File: src/librustc_trans/back/symbol_export.rs (hunk 2 above)
Comment (michaelwoerister) [inline src/librustc_trans/back/symbol_export.rs]: OK

File: src/librustc_trans/back/lto.rs (hunk 3 above)
Comment (michaelwoerister) [inline src/librustc_trans/back/lto.rs]: Will fix it.

File: src/librustc_trans/back/symbol_export.rs (hunk 2 above)
Comment (michaelwoerister) [inline src/librustc_trans/back/symbol_export.rs]: So it is more or less like a cdylib, right? Or a Rust dylib?

**Comment by michaelwoerister** [comment]
The main changes in here are:

1. Every symbol that is not exported is declared as `hidden` in LLVM in the `internalize_symbols` pass. Whether a symbol is exported is determined the following way:
    1. Determine the "export threshold" of this compilation session. If we are only compiling cdylibs or staticlibs or executables, this threshold is `export just #[no_mangle] extern "non-Rust"`. If there is a Rust dylib in the mix, the threshold is `export anything that is monomorphic, non-inline, and publicly reachable`
    2. Walk all declarations and definitions from this crate, internalize if possible. Failing that, mark as `hidden` if below the threshold.

2. The new `ExportedSymbols` data structure computes and stores a `SymbolExportLevel` for each symbol and makes this available to the code that generates the export list for a given linker invocation. This way, we can generate a tighter export list for a cdylib even though we could not declare things as hidden in the `internalize_symbols` pass.

3. Use `--version-script` instead of `--retain-symbols` when invoking the linker. This just seems to be some kind of random bug that would lead to our export lists always being ignored by the linker.

I'm just thinking, maybe it would be sufficient to just always rely on explicit export lists and don't bother with marking things as hidden in LLVM. I'm not sure though if LLVM can already do some optimizations based on the fact that something is `hidden`.

**Comment by froydnj** [comment]
> I'm just thinking, maybe it would be sufficient to just always rely on explicit export lists and don't bother with marking things as hidden in LLVM. I'm not sure though if LLVM can already do some optimizations based on the fact that something is hidden.

More information for the compiler is almost always better.  I think the roughly equivalent situation in C/C++ on Linux is using `--version-script` at link time for a shared library rather than passing `-fvisibility=hidden` and explicitly specifying in the source code the visibility of symbols that you want to export: you'll wind up with the same exported symbols in both cases, but the compiler can generate better code if it knows that you'll never export a given symbol, since you don't have to go through the PLT.

File: src/librustc_trans/back/symbol_export.rs (hunk 2 above)
Comment (alexcrichton) [inline src/librustc_trans/back/symbol_export.rs]: Yeah, more like a cdylib

**Comment by alexcrichton** [comment]
Yeah I agree with @froydnj that getting info in the compiler is best. We also want to help out the staticlib case as much as possible where we don't always control the linker invocation.

@michaelwoerister do you have thoughts on the ["rlib in dylib case" above](https://github.com/rust-lang/rust/pull/38117#pullrequestreview-11061636)

**Comment by michaelwoerister** [comment]
I tried to list the various cases in the [test](https://github.com/rust-lang/rust/pull/38117/commits/46594c8dd2a2f23d7888b196aa5dcfb8581501ea#diff-5614d27cefadb5c1c3ac8d92682a43ccR9).

To summarize:
- A Rust dylib will export everything from any rlibs it includes
- A cdylib will only export public, non-Rust, `#[no_mangle]` symbols from any rlibs it includes.

So this is pretty much what you said, I think.

As for static libs: They inherit whatever we mark things as in LLVM, I'm not sure actually if the version script will cause the linker to put things not in there into the local symbol table, even if not building a dynamic library. Ideally we'd only want things in the staticlib's global symbol table that would also be exported from a cdylib, right?

**Comment by alexcrichton** [comment]
Yeah in theory a staticlib exported symbol list is precisely the same as a cdylib. Unfortunately though, as you've seen, I don't think we can actually do that. Maybe one day we can emit something that's like "pass these extra args to the linker" but we're not quite there just yet.

**Comment by michaelwoerister** [comment]
This and what @froydnj mentioned would be good reasons for switching to rlibs that don't contain any machine code. Then we could control all of this at the leaf products by just declaring things as hidden in LLVM.

**Comment by cuviper** [comment]
> A Rust dylib will export everything from any rlibs it includes

This is a conservative policy, but not necessarily optimal, right?  e.g. A dylib which was linked with the `std` rlib probably doesn't need to export *everything* from `std`.

**Comment by michaelwoerister** [comment]
> A dylib which was linked with the std rlib probably doesn't need to export everything from std.

Everything that is publicly reachable, not marked with `#[inline]` and not generic. But yeah, that's still a lot. But it's also what our rules currently say.

Rust dylibs are strange things anyway. I'm wondering if we should get rid of them, or at least declare them as unstable, not that we have cdylibs.

**Comment by alexcrichton** [comment]
> switching to rlibs that don't contain any machine code

Hey if we get a fast enough code generator I'm all for this, it'd solve quite a few problems.

**Comment by cuviper** [comment]
For instance, if my library was a `FancyVec`, then I'd expect it to export `std` stuff reachable through the `Vec` interface I'm wrapping.  But ideally it would not export anything from unrelated parts of `std`.

**Comment by michaelwoerister** [comment]
@cuviper If `FancyVec` was part of a cdylib, then nothing at all from std would be re-exported. But since `FancyVec` is probably a generic type, there is no benefit in putting it into a dynamic library at all. If you put it into an rlib, then you don't have to worry about symbol visibility.

**Comment by cuviper** [comment]
@michaelwoerister There still could be some benefit to a dylib, if there's some part of that fancy algorithm that doesn't depend on the generic type, perhaps only dealing with the length.

But OK, the example doesn't have to be generic, so let's try something concrete.  Consider `num::BigUint`, which wraps a `Vec<u32>`.  Compiled as a dylib, I get a lot of dynamic symbols which are clearly unrelated, like stuff in `std::net`.

I'm not suggesting this needs to block the PR, just that there's further room for improvement.  In practice people should favor rlibs anyway, since there's no stable ABI in a dylib.

**Comment by michaelwoerister** [comment]
@cuviper But how does the compiler know what is related and what isn't? Maybe you included some rlib only because you want to make it's symbols available in your dylib? But I agree that this is not exactly a desirable state of affairs.

**Comment by alexcrichton** [comment]
@cuviper note that symbol visibility was a primary reason cdylibs were created. If you want strict control over symbols then dylib isn't the crate type you want (due to the way rustc uses it). A cdylib, however, is intended to be as conservative as possible with symbol exports.

**Comment by cuviper** [comment]
@michaelwoerister 
> Maybe you included some rlib only because you want to make it's symbols available in your dylib?

Could be, but I think that's a niche case.  Maybe there could be a `#[link]` or something on the crate import to explicitly support that.

@alexcrichton 
> If you want strict control over symbols then dylib isn't the crate type you want

OK, but I'm saying the "uncontrolled" state of symbols is not ideal.  I think the default should only export things the crate itself makes directly reachable.  I might want `num` in a dylib (because reasons), but it's silly for that dylib to carry code for `std::net` at all, let alone export it.

Maybe we should move this to its own issue.  P-low because dylibs are a bit silly anyway.

**Comment by vadimcn** [comment]
How important is it to allow symbol preemption in dylibs?  This feature does not exist on Windows and I can't say I ever missed it.  
Could we indeed just mark all symbols as "hidden" or "private" in LLVM IR and use linker scripts to export only the ones we need?

**Comment by michaelwoerister** [comment]
> Could we indeed just mark all symbols as "hidden" or "private" in LLVM IR and use linker scripts to export only the ones we need?

Does anybody how that would interact with codegen? If LLVM relies on the symbol not being exported but then the linker exports it anyway, that sounds like a potential problem.

**Comment by nagisa** [comment]
LLVM will outright remove code for provably unreachable stuff.

**Comment by cuviper** [comment]
Some platforms have differences in calling convention too -- e.g. IIRC ppc64le can have multiple entry points depending one whether you need to setup the TOC pointer.

**Comment by vadimcn** [comment]
> Does anybody how that would interact with codegen? If LLVM relies on the symbol not being exported but then the linker exports it anyway, that sounds like a potential problem.

If I am reading [this](http://www.airs.com/blog/archives/307) correctly, hidden symbols are still available for static linking, however they are not put into the dynamic symbol table, should that object be linked into a shared lib.  Also, internal references to them inside shared lib will not use indirection via PLT.

**Comment by michaelwoerister** [comment]
> LLVM will outright remove code for provably unreachable stuff.

I don't think that `hidden` things become unreachable, they are just not in the dynamic symbol table (for dylibs) / go into the local symbol table instead of the global one (for object files/static libs).

> Some platforms have differences in calling convention too -- e.g. IIRC ppc64le can have multiple entry points depending one whether you need to setup the TOC pointer.

That sounds like a problem.

**Comment by alexcrichton** [comment]
@michaelwoerister 

btw the only changes wrt this PR itself I see before r+ are:

* The `symbol-visibility` test is likely to fail on non-Linux
* The proc-macro crate type can be more restrictive

(just to summarize)

**Comment by michaelwoerister** [comment]
> The symbol-visibility test is likely to fail on non-Linux

This should work now for OSX and MSVC. On MINGW I see too many symbols included for cdylibs, so there's still a bug in there somewhere.

> The proc-macro crate type can be more restrictive

This should be fixed by https://github.com/rust-lang/rust/pull/38117/commits/581c29284b23823be25f700052b966efe1b65894 now.

**Comment by alexcrichton** [comment]
r=me whenever this passes tests, or also feel free to use bors to see if it passes tests :)

**Comment by michaelwoerister** [comment]
I disabled tests on MINGW now. The linker there seems to ignore `--version-script` and `--dynamic-symbols`. It does *something* when `--retain-symbols` is passed, that is, afterwards `nm` will only list the symbols specified. But `dumpbin`, which I think is more authoritative on Windows, still lists all symbols under `/EXPORTS` (@retep998, any clue what that is about?). So on MINGW you get the old behavior basically.

**Comment by alexcrichton** [comment]
@michaelwoerister sounds ok to me, we can always look more into it if it becomes a problem.

**Comment by m4b** [comment]
Have you tried `--export-all-symbols` in combination with the version script?

You might also try messing with `--ouput-def` and then using it as input to the linker (on mingw windows)

`nm` is probably printing symbol information from the COFF portion of the binary and not the Microsoft PE extended part, so yea I'd default to ms binutils in this case

**Comment by michaelwoerister** [comment]
@bors r=alexcrichton

Let's give it a try `:)`

**Pull request merged**